### PR TITLE
CachingReader: Reduce log spam for corrupt files

### PR DESCRIPTION
**Pull request by uklotzde**
Partially fixes (decoding errors at the end of a file): https://bugs.launchpad.net/mixxx/+bug/1470010
- Shorten readable length of corrupt files to avoid repeated decoding
- Code cleanups: Consistent naming, remove unncessary checks and local variables, ...

### CODE REVIEW

File: src/cachingreader.cpp [hunk 1]
```diff
@@ -11,6 +11,20 @@
 #include "util/math.h"
 #include "util/assert.h"
 
+namespace {
+
+// To prevent every bit of code having to guess how many samples
+// forward it makes sense to keep in memory, the hinter can provide
+// either 0 for a forward hint or -1 for a backward hint. We should
+// be calculating an appropriate number of samples to go backward as
+// some function of the latency, but for now just leave this as a
```
Comment (daschuer) [inline src/cachingreader.cpp]: some time ago we have changed the wording from "latency" to "audio buffer". 
Did you mean here: "as some function of the selected audio buffer in hardware preferences?"

File: src/cachingreader.cpp [hunk 2]
```diff
@@ -11,6 +11,20 @@
...
+// constant. 2048 is a pretty good number of samples because 25ms
```
Comment (daschuer) [inline src/cachingreader.cpp]: Why 25 ms? Default is 11.6 ms.

Comment (daschuer) [inline src/cachingreader.cpp]: Ah, never mid. I have just noticed you have just copied the original comment. So fixing this issue is not part of this PR.

**Comment by daschuer** [comment]
I can confirm that the log spam is reduced playing the faulty files from 
https://github.com/jrmuizel/AudioScan-opus

However now I have files that won't end. They are stucked just before the and with schakin waveforms
https://github.com/jrmuizel/AudioScan-opus/blob/master/t/mp3/v2.3-iso-8859-1.mp3
This does not happen without the patch. 

This file also prevent AutoDJ form continuing.

**Comment by uklotzde** [comment]
Thanks for finding another edge case, Daniel :+1:

**Comment by uklotzde** [comment]
My last commit should fix this. Now the requested read buffer is simply 
filled with zeros if sample data is unreadable. This applies to both the 
beginning ("preroll") as well as the end of the file.

**Comment by daschuer** [comment]
Issue fixed. 

However I found crasher: 
https://github.com/jrmuizel/AudioScan-opus/blob/master/t/mp4/array-keys-int.m4a
ReadAtom: "/home/daniel/workspace/t/libaudio-scan-perl-master/t/mp4/array-keys-int.m4a": invalid atom size, extends outside parent atom - skipping to end of "" "mdat" 5848240 vs 135015
GetTrackMediaDataName: "/home/daniel/workspace/t/libaudio-scan-perl-master/t/mp4/array-keys-int.m4a": track 1 has more than 1 child atoms in stsd
```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffb89fd700 (LWP 30058)]
__strcasecmp_l_ssse3 () at ../sysdeps/x86_64/multiarch/../strcmp.S:209
209	../sysdeps/x86_64/multiarch/../strcmp.S: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  __strcasecmp_l_ssse3 () at ../sysdeps/x86_64/multiarch/../strcmp.S:209
#1  0x00007fffd1b1c14c in findFirstAudioTrackId (hFile=0x7fffb4002100)
    at plugins/soundsourcem4a/soundsourcem4a.cpp:72
#2  Mixxx::SoundSourceM4A::tryOpen (this=0x7fffb4003ca0, audioSrcCfg=...)
    at plugins/soundsourcem4a/soundsourcem4a.cpp:123
#3  0x0000000000ac646e in Mixxx::SoundSource::open (this=0x7fffb4003ca0, 
    audioSrcCfg=...) at src/sources/soundsource.cpp:28
#4  0x0000000000ac4865 in SoundSourceProxy::openAudioSource (
    this=this@entry=0x7fffb89fcd00, audioSrcCfg=...)
    at src/soundsourceproxy.cpp:152
#5  0x00000000004de864 in openAudioSourceForReading (audioSrcCfg=..., 
    pTrack=...) at src/cachingreaderworker.cpp:167
#6  CachingReaderWorker::loadTrack (this=this@entry=0x1e6fb18, pTrack=...)
    at src/cachingreaderworker.cpp:199
#7  0x00000000004dfb2b in CachingReaderWorker::run (this=0x1e6fb18)
    at src/cachingreaderworker.cpp:150
#8  0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#9  0x00007ffff31e6182 in start_thread (arg=0x7fffb89fd700)
    at pthread_create.c:312
#10 0x00007ffff189247d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 
```

**Comment by uklotzde** [comment]
We have already fixed this one some time ago. The current master (just did a 
clean build) doesn't crash on this file. I get an appropriate error log and 
Mixxx refuses to load the file:

Debug [Main]: WCoverArt::slotCoverFound WCoverArt(0x4bee4d0) 
"CoverInfo(METADATA,GUESSED,,48529,/home/uk/Music/Corrupt Tracks 
(Link)/array-keys-int.m4a)" QSize(300, 300)
Read: "/home/uk/Music/Corrupt Tracks (Link)/array-keys-int.m4a": stsd 
inconsistency with number of entries
ReadAtom: "/home/uk/Music/Corrupt Tracks (Link)/array-keys-int.m4a": invalid 
atom size, extends outside parent atom - skipping to end of "" "mdat" 
5848240 vs 135015
GetTrackMediaDataName: "/home/uk/Music/Corrupt Tracks 
(Link)/array-keys-int.m4a": track 1 has more than 1 child atoms in stsd
Warning [CachingReaderWorker 1]: No AAC track found: 
"file:///home/uk/Music/Corrupt Tracks (Link)/array-keys-int.m4a"
Warning [CachingReaderWorker 1]: Failed to open SoundSource
Warning [CachingReaderWorker 1]: Failed to open file: 
"/home/uk/Music/Corrupt Tracks (Link)/array-keys-int.m4a"
Debug [CachingReaderWorker 1]: "[Channel1]" CachingReaderWorker::loadTrack() 
load failed for" "/home/uk/Music/Corrupt Tracks (Link)/array-keys-int.m4a" 
", file invalid, unlocked reader lock
Debug [Main]: Failed to load track "/home/uk/Music/Corrupt Tracks 
(Link)/array-keys-int.m4a" "The file '/home/uk/Music/Corrupt Tracks 
(Link)/array-keys-int.m4a' could not be loaded."

**Comment by uklotzde** [comment]
2015-06-30 Merge pull request #626 from 
uklotzde/CachingReaderAvoidUnnecessarySeeking

Fix outstanding SoundSource issues

**Comment by daschuer** [comment]
My test result was generated by this PR merged to the current master. 
So https://github.com/mixxxdj/mixxx/pull/626 should be part of it. 
Maybe the issue is lightly different or I have messed up my repository.

**Comment by daschuer** [comment]
Ah, sorry for the false alarm, I have missed to use the recent m4a plugin from gdb. 
Now it works.

**Comment by uklotzde** [comment]
There is a bug in the preroll calculation that causes segmentation faults during the tests. Fix will follow.

File: src/cachingreader.cpp [hunk 3]
```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
...
-    int samples_remaining = num_samples;
+    SINT samplesRead = 0;
 
+    SINT frameIndex = samples2frames(sample);
+    SINT numFrames = samples2frames(numSamples);
 
     // TODO: is it possible to move this code out of caching reader
     // and into enginebuffer?  It doesn't quite make sense here, although
     // it makes preroll completely transparent to the rest of the code
     // if we're in preroll...
```
Comment (daschuer) [inline src/cachingreader.cpp]: It is not the first time I am struggling with this comment. I like the current solution, because it simplifies the code elsewhere. If you think the same, we should remove this comment.

File: src/cachingreader.cpp [hunk 4]
```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
...
-    if (sample < 0) {
-        int zero_samples = math_min(-sample, samples_remaining);
-        DEBUG_ASSERT(0 <= zero_samples);
-        DEBUG_ASSERT(zero_samples <= samples_remaining);
-        SampleUtil::clear(buffer, zero_samples);
-        samples_remaining -= zero_samples;
-        if (samples_remaining == 0) {
-            //everything is zeros, easy
-            return zero_samples;
+    if (Mixxx::AudioSource::getMinFrameIndex() > frameIndex) {
```
Comment (daschuer) [inline src/cachingreader.cpp]: Please add a comment like this: 
// frame index points to the pre-roll, a region before the first track sample

File: src/cachingreader.cpp [hunk 5]
```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
...
+        const SINT prerollFrames = math_min(numFrames,
+                Mixxx::AudioSource::getMinFrameIndex() - frameIndex);
+        const SINT prerollSamples = frames2samples(prerollFrames);
+        SampleUtil::clear(buffer, prerollSamples);
+        if (prerollSamples == numSamples) {
+            return numSamples; // early exit
```
Comment (daschuer) [inline src/cachingreader.cpp]: The comment explains the keyword return. ;-) What about: 
// All requests samples are in the pre-roll: done!

File: src/cachingreader.cpp [hunk 6]
```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
...
-        // If we did not decide to read any samples from this chunk then that
-        // means we have exhausted all the samples in the song.
-        if (frames_to_read == 0) {
-            break;
-        }
+            // The max. readable frame index might change with each
+            // read operation!
+            const SINT maxReadableFrameIndex = math_min(maxFrameIndex, m_maxReadableFrameIndex);
+            if (maxReadableFrameIndex <= frameIndex) {
+                break; // exit loop
```
Comment (daschuer) [inline src/cachingreader.cpp]: // exit loop   Sure ;-)

Comment (daschuer) [inline src/cachingreader.cpp]: // exit loop to fill remaining buffer with silence

**Comment by uklotzde** [comment]
Couldn't resist to an OO refactoring of the sensible CachingReader stuff. Still not perfect, but hopefully much cleaner and safer than before. The code is guarded by lots of debug assertions to detect any inconsistencies in the handover of chunks between cache and worker early during development.

As a bonus I added a simple strategy to compensate for seek inaccuracies in faulty files.

File: src/cachingreaderchunk.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    m_state = FREE;
+}
+
+void CachingReaderChunk::insertIntoListBefore(
+        CachingReaderChunk* pBefore) {
+    DEBUG_ASSERT(nullptr == m_pNext);
+    DEBUG_ASSERT(nullptr == m_pPrev);
+
+    m_pNext = pBefore;
+    if (pBefore) {
```
Comment (daschuer) [inline src/cachingreaderchunk.cpp]: I think this can also be an assertion.

File: src/cachingreaderchunk.cpp [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        if (pBefore->m_pPrev) {
+            m_pPrev = pBefore->m_pPrev;
+            DEBUG_ASSERT(m_pPrev->m_pNext == pBefore);
+            m_pPrev->m_pNext = this;
+        }
+        pBefore->m_pPrev = this;
+    }
```
Comment (daschuer) [inline src/cachingreaderchunk.cpp]: Do we need to adjust the ppTail here?

File: src/cachingreaderchunk.cpp [hunk 9]
```diff
@@ -0,0 +1,186 @@
+#include <QtDebug>
+
+#include "cachingreaderchunk.h"
+
+#include "sampleutil.h"
+#include "util/math.h"
+
+
+const SINT CachingReaderChunkWorker::kDefaultIndex = -1;
```
Comment (daschuer) [inline src/cachingreaderchunk.cpp]: I think this should be kInvalidIndex

File: src/cachingreaderchunk.cpp (hunk 8 above)
Comment (uklotzde) [inline src/cachingreaderchunk.cpp]: Nope, because we insert the new element before pBefore and the tail element 
will remain the same. Only the head may change if pBefore is currently the 
head of the queue, but only the caller knows this.

I think about how to unify these two operations.

File: src/cachingreaderchunk.cpp (hunk 7 above)
Comment (uklotzde) [inline src/cachingreaderchunk.cpp]: To insert an element at the head of the queue you need to pass a pointer to 
the head pHead == pBefore. If the list is currently empty there is no head 
and consequently pHead == pBefore == nullptr. If we add an assertion the 
caller would be responsible to check for emptiness before calling this 
function. I prefer to handle the special case of an empty list inside this 
function.

File: src/cachingreaderchunk.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    // The worker thread will fill the sample buffer and
+    // set the frame count.
+    CSAMPLE* const m_sampleBuffer;
+    volatile SINT m_frameCount;
+};
+
+// The derived class is only accessible for the cache, but not the
+// worker thread. The state READ_PENDING indicates that the worker
+// thread is in control.
+class CachingReaderChunk: public CachingReaderChunkWorker {
```
Comment (daschuer) [inline src/cachingreaderchunk.h]: The current state is a bit confusing, because first I thought CachingReaderChunkWorker is a thread. 
I think I got why you split up the class: to separate the function that are accessed from different threads. 
So it is probably only a naming issue. 

C++ does not enforce this restrictions yet, right?  Maybe we can manage it by accessing the chunks via two different
abstract classes. Like ICachinReaderChunkForWorker  and ICachinReaderChunkForCache

File: src/cachingreaderworker.cpp [hunk 11]
```diff
@@ -1,122 +1,66 @@
...
-    // Initialize the output parameter
+    // Initialize the output parameters
+    update->maxReadableFrameIndex = m_maxReadableFrameIndex;
     update->chunk = request->chunk;
-    update->chunk->frameCountRead = 0;
-    update->chunk->frameCountTotal = 0;
 
-    const int chunk_number = request->chunk->chunk_number;
-    if (!m_pAudioSource || chunk_number < 0) {
+    if (!update->chunk->isReadable(m_pAudioSource, m_maxReadableFrameIndex)) {
```
Comment (daschuer) [inline src/cachingreaderworker.cpp]: This check might be redundant. 
Can't we integrate this and the isReadable call below into readSampleFrames?

File: src/cachingreaderchunk.h (hunk 10 above)
Comment (uklotzde) [inline src/cachingreaderchunk.h]: I've just split a reduced API for the worker by separating it into a base 
class. It was a compromise, but we can easily separate those classes even 
further into CommonBase/...ForWorker/...ForCache. I think pure interfaces 
without any members/implementation would be overkill.

Cross-casting from ...ForWorker from/to ...ForCache might look a little 
nasty. Now only a single downcast in the cache is needed. We could pass a 
pointer to the common base class in the communication between cache and 
worker. Then we only need 2 downcasts instead of weird cross-casts.

File: src/cachingreaderworker.cpp [hunk 12]
```diff
@@ -1,122 +1,66 @@
...
-        ChunkReadRequest* request,
+void CachingReaderWorker::processCachingReaderChunkReadRequest(
+        CachingReaderChunkReadRequest* request,
         ReaderStatusUpdate* update) {
-    //qDebug() << "Processing ChunkReadRequest for" << chunk_number;
+    //qDebug() << "Processing CachingReaderChunkReadRequest for" << chunk_number;
 
-    // Initialize the output parameter
+    // Initialize the output parameters
+    update->maxReadableFrameIndex = m_maxReadableFrameIndex;
```
Comment (daschuer) [inline src/cachingreaderworker.cpp]: Since the m_maxReadableFrameIndex may be changed below, is it line here corrector? 
maybe we can entirely ditch m_maxReadableFrameIndex and use always directly the m_pAudioSource value.

File: src/cachingreaderchunk.h [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        READ_PENDING
+    };
+
+    State getState() const {
+        return m_state;
+    }
+
+    // Please note that the state must exclusively be controlled by the
+    // cache and not the worker thread!
+    void beginReading() {
```
Comment (daschuer) [inline src/cachingreaderchunk.h]: The name may indicate that the worker is in charge to call this and finishReading below. 
Maybe we find a fancy name like:
markForReading() or enableForReading() or just setReadPending(); 
finishReading() may become chunkReceived() or something.

Comment (daschuer) [inline src/cachingreaderchunk.h]: Or
releaseForWorker()
aquireFromWorker()

File: src/cachingreaderworker.cpp (hunk 12 above)
Comment (uklotzde) [inline src/cachingreaderworker.cpp]: I've added more comments to the worker that hopefully explain why we need 
two invocations of isReadable(), one before and one after reading data from 
the audio source to determine the result status of the read request.

File: src/cachingreaderworker.cpp (hunk 11 above)
Comment (uklotzde) [inline src/cachingreaderworker.cpp]: The isReadable() function introduces some redundancy with a negligible 
performance overhead. On the other hand it saves us from defining an 
additional enum and type that encapsulates the complete read result that is 
needed by the worker.

File: src/cachingreaderchunk.h [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        READY,
+        READ_PENDING
+    };
+
+    State getState() const {
+        return m_state;
+    }
+
+    // The state is controlled by the cache as the owner of each chunk!
+    void giveToWorker() {
```
Comment (daschuer) [inline src/cachingreaderchunk.h]: Nice! :-)

File: src/cachingreaderworker.cpp (hunk 12 above)
Comment (daschuer) [inline src/cachingreaderworker.cpp]: In the case that m_maxReadableFrameIndex is reduced by this call
```
const SINT framesRead = request->chunk->readSampleFrames(
+            m_pAudioSource, &m_maxReadableFrameIndex);
```
we report an outdated update->maxReadableFrameIndex to the reader. 
Is it desired?

Comment (uklotzde) [inline src/cachingreaderworker.cpp]: Good catch, I just pushed a commit with the missing update. It didn't harm, 
but the fix might prevent a repeated invalid read from a corrupt file.

**Comment by daschuer** [comment]
Sorry, the CachingReaderChunk split still not works for me. 

For example: CachingReaderChunkForWorker::indexForFrame is accessed from cachingReader.

In the current state it looks more confusing than helpful to me. 

A better solution may look like this, using two friend Keys: 
http://stackoverflow.com/questions/1609472/friend-class-with-limited-access

Since the surrounding code is (thanks to all the refactoring and comments) pretty clear now, 
there is IMHO no need for such a split or any extra effort. 
So what about renaming just CachingReaderChunkForWorker to CachingReaderChunkBase change the comments accordingly and merge this PR? 

By the way: The rest looks good to me. Thank you very much.

**Comment by uklotzde** [comment]
I didn't take the static functions and constants in the base class into 
account while renaming. Of course, that looks confusing ;)

I will split the chunk into 3 classes:
* CachingReaderChunk
* CachingReaderChunkForOwner (i.e. the cache)
* CachingReaderChunkForWorker (i.e. the worker thread)

**Comment by uklotzde** [comment]
Ups, no, 3 classes don't work without further complications. The following 
two classes are a viable solution, both comprehensive and effective:

* CachingReaderChunk (base class for both the owner and the worker)
* CachingReaderChunkForOwner (i.e. the cache)

**Comment by daschuer** [comment]
OK, this will work for me as well.

**Comment by daschuer** [comment]
LGTM Ready to merge?

**Comment by uklotzde** [comment]
Ready. Further improvements can be done later.

**Comment by daschuer** [comment]
Fine :-) Thank you for another nice PR.

**Comment by uklotzde** [comment]
Thanks for the review, Daniel! Always a pleasure :)
